pkg/wakaama/include/lwm2m_objects/device.h

```diff
@@ -0,0 +1,120 @@
+    LWM2M_RES_FRESET,           /**< Perform a factory reset of the device */
+    LWM2M_RES_POWER_SRC,        /**< Available power sources */
+    LWM2M_RES_POWER_VOL,        /**< Present voltage for each power source */
+    LWM2M_RES_POWER_AMP,        /**< Present current for each power source */
+    LWM2M_RES_BATTERY_LEVEL,    /**< Current battery level as a percentage */
+    LWM2M_RES_MEM_FREE,         /**< Estimated current available storage (kB) */
+    LWM2M_RES_ERROR_CODE,       /**< Last error code */
+    LWM2M_RES_ERROR_CODE_RESET, /**< Delete all error code instances */
+    LWM2M_RES_TIME,             /**< Current UNIX time of the client */
+    LWM2M_RES_TIME_OFFSET,      /**< Indicated the UTC offset for the device */
+    LWM2M_RES_TIME_ZONE,        /**< Indicates the time zone of the device */
+    LWM2M_RES_BINDINGS,         /**< Indicates supported bindings and modes on the client */
+    LWM2M_RES_TYPE,             /**< Type of device */
+    LWM2M_RES_HW_VERSION,       /**< Current hardware version of the device */
+    LWM2M_RES_SW_VERSION,       /**< Current software version on the device */
+    LWM2M_RES_BATTERY_STATUS,   /**< Battery status when internal battery is present */
+    LWM2M_RES_MEM_TOTAL,        /**< Total amount of storage space in the device (kB*/
+    LWM2M_RES_EXT_DEV_INFO,     /**< External device object instance */
+    LWM2M_DEVICE_RESOURCES      /**< Number of resources */
+};
+
+/**
+ * @brief Error codes for the
+ * @ref lwm2m_device_resources::LWM2M_RES_ERROR_CODE "Error" resource in the
+ * device object of LwM2M
+ */
+enum lwm2m_device_error_codes {
+    LWM2M_DEVICE_ERR_NO_ERR     = 0, /**< No error */
+    LWM2M_DEVICE_ERR_LOW_BATT   = 1, /**< Low battery power */
+    LWM2M_DEVICE_ERR_EXT_OFF    = 2, /**< External power supply off */
+    LWM2M_DEVICE_ERR_GPS_ERR    = 3, /**< GPS module failure */
+    LWM2M_DEVICE_ERR_LOW_SIGNAL = 4, /**< Low received signal strength */
+    LWM2M_DEVICE_ERR_NO_MEM     = 5, /**< Out of memory */
+    LWM2M_DEVICE_ERR_SMS_ERR    = 6, /**< SMS failure */
+    LWM2M_DEVICE_ERR_IP_ERR     = 7, /**< IP connectivity failure */
+    LWM2M_DEVICE_ERR_PERIPH_ERR = 8  /**< Peripheral malfunction */
+};
+
+/**
+ * @brief Descriptor of a LwM2M device object instance
+ */
+typedef struct {
+    uint8_t *power_sources;      /**< types of power sources (0-7) */
+    uint16_t *power_voltage;     /**< voltage of power sources in mV */
+    uint16_t *power_current;     /**< current of power sources in mA */
+    uint8_t battery_status;      /**< battery status (0-6) */
+    uint32_t mem_total;          /**< amount of memory on the device in kB */
+    uint16_t(*ext_dev_info)[2];  /**< external devices information */
+    uint8_t ext_dev_info_len;    /**< amount of external devices information */
+    uint8_t error_code[7];       /**< error codes */
+    uint8_t error_code_used;     /**< amount of error codes used */
+} dev_data_t;
```

Comment:
Does it make sense to hide this in `device.c` since it is not used outside? Is the plan to eventually make this visible externally?